### PR TITLE
[FEAT] Add File to Context Automatically When No Text in Box

### PR DESCRIPTION
**Pull request by charlwillia6**
## Description ✏️

Added automatic context pinning when the chat box is empty for the current file being edited.  It automatically switches context when switching between files.  

image
image

## Checklist ✅

- [x] I have added screenshots (if UI changes are present).
- [x] I have done a self-review of my code.
- [x] I have manually tested my code (if applicable).

### CODE REVIEW

File: gui/src/components/mainInput/TipTapEditor.tsx [hunk 1]
```diff
@@ -132,7 +130,7 @@ const getPlaceholder = (historyLength: number, location: any) => {
     : "Ask a follow-up";
 };
 
-function getDataUrlForFile(file: File, img): string {
+const getDataUrlForFile = (file: File, img): string => {
```
Comment (Fryingpannn) [inline gui/src/components/mainInput/TipTapEditor.tsx]: I know the urge to change the `function` to arrow functions, but let's avoid doing this x) We should minimize merge conflicts with upstream, so unfortunately it is worth the practice sacrifice.

File: gui/src/redux/slices/stateSlice.ts [hunk 2]
```diff
@@ -270,10 +272,9 @@ export const stateSlice = createSlice({
           : [{ ...item, editing: true }];
         return;
       }
-      // TODO
     },
     addContextItems: (state, action: PayloadAction<ContextItemWithId[]>) => {
-      state.contextItems = state.contextItems.concat(action.payload);
+      state.contextItems.push(...action.payload);
```
Comment (Fryingpannn) [inline gui/src/redux/slices/stateSlice.ts]: Not sure if we should use `push` here, I think for state we want things to be immutable, so should return a new object like `concat` does.

**Comment by Fryingpannn** [comment]
Tested. Current behaviour:

- Focused editor file will appear mentioned in chat if there's nothing in the current chat
- If there's already codeblock, slashcommand or text, then won't put current open file in chat.
- Will swap current file if no other content is in the text box upon switching files
- If close current file, the file mention still remains in chat (non-ideal)
- If have content in chat, the current mention file will not change

File: extensions/vscode/src/extension/VsCodeExtension.ts [hunk 3]
```diff
@@ -360,15 +361,23 @@ export class VsCodeExtension {
...
+    context.subscriptions.push(
+      vscode.window.onDidChangeTextEditorSelection((event) => {
+        const filepath = event.textEditor?.document?.uri.fsPath || null;
+        this.debouncedNotifyEditorChange(filepath);
+      })
+    );
+
     this.ide.onDidChangeActiveTextEditor((filepath) => {
-      this.core.invoke("didChangeActiveTextEditor", { filepath });
+      this.debouncedNotifyEditorChange(filepath || null);
```
Comment (Fryingpannn) [inline extensions/vscode/src/extension/VsCodeExtension.ts]: I wonder if this debouncing is needed, also with 100ms, I don't think anyone is switching that fast between tabs that it's worth it. This maybe more for typing. I'd say no need debounce here unless you feel like there are other factors that deterioriate the ux without it.

**Comment by Himanshu-Singh-Chauhan** [comment]
damn, this is big, took time to review.
awesome work @charlwillia6 
i tried to show "current file" text on the mention, didnt work, it does not send the file contents.
Also, I feel including current file in every chat is fine, cause user may make changes and when asks ai next step, ai can read the modified file again and see what changes were applied and what were not. 
image

File: extensions/vscode/src/extension/VsCodeExtension.ts [hunk 4]
```diff
@@ -360,15 +361,23 @@ export class VsCodeExtension {
...
         documentContentProvider,
       ),
     );
 
+    context.subscriptions.push(
+      vscode.window.onDidChangeTextEditorSelection((event) => {
+        const filepath = event.textEditor?.document?.uri.fsPath || null;
+        this.debouncedNotifyEditorChange(filepath);
+      })
+    );
```
Comment (Himanshu-Singh-Chauhan) [inline extensions/vscode/src/extension/VsCodeExtension.ts]: what is this event? is this when we select text in a file. if so, then maybe we don't want to trigger notifyeditorchange

File: gui/src/components/mainInput/getSuggestion.ts [hunk 5]
```diff
@@ -21,37 +21,48 @@ function getSuggestion(
...
 
       return {
         onStart: (props) => {
-          component = new ReactRenderer(MentionList, {
-            props: { ...props, enterSubmenu, onClose: onExit },
-            editor: props.editor,
+          requestAnimationFrame(() => {
+            component = new ReactRenderer(MentionList, {
+              props: { 
+                ...props,
```
Comment (Himanshu-Singh-Chauhan) [inline gui/src/components/mainInput/getSuggestion.ts:36]: whats this changes, mmm, i never saw tippy tooltip, where does this show?

Comment (nang-dev) [reply gui/src/components/mainInput/getSuggestion.ts:36]: Hmmm

**Comment by nang-dev** [comment]
i added a pull from main . this is good so far! thx sm

2 bugs i see:
1. "New chat" should include the current file
Screenshot 2024-11-02 at 1 08 50 PM

2. All subsequent chats in any thread should have ALL of this functionality turned off / ignored.
Screenshot 2024-11-02 at 1 09 22 PM

nice to have: if the context used can have page.tsx (Active File)
Screenshot 2024-11-02 at 1 07 48 PM

**Comment by charlwillia6** [comment]
@nang-dev @Fryingpannn @Himanshu-Singh-Chauhan - New changes have been pushed.  Please review and test.

File: gui/src/components/mainInput/getSuggestion.ts (hunk 5 above)
Comment (charlwillia6) [reply gui/src/components/mainInput/getSuggestion.ts:36]: It looks like this maybe a refactoring change (not on purpose).  If you look though, it was there previously.  Scroll down and look at the diff.

File: extensions/vscode/src/extension/VsCodeExtension.ts (hunk 4 above)
Comment (charlwillia6) [reply extensions/vscode/src/extension/VsCodeExtension.ts]: This even is specific to getting the file path of the selected file.

**Comment by charlwillia6** [comment]
This has been tested, and with @Himanshu-Singh-Chauhan changes, this seems to work just as requested.  I also merged main into this PR.

**Review by Fryingpannn — APPROVED** [review]

**Review by Fryingpannn — APPROVED** [review]

**Comment by Fryingpannn** [comment]
Gj Charles.

Leaving some comments for future.

Notice
- If user erases context and uses their own context, then switch file, it removes the context they manually added. This is because the feature works by checking if only 1 context file, then if user switch files, also switch that context file. No distinction is made between a default active file and a file that user added, only the count. However, I think this is an obscure edge enough that we can let it slide for now. Until we have the polished active file UI in chatbox.
- for `slashCommand`s, this is currently also erased when user switches file and user only has a slash command with no text in chatbox. I think this is because it's not checking the slash command type, if already existing then should not replace it (if there's text, it won't replace). Will also let this slide for now since usually when a user uses slash command, I think not likely they will change files without even typing a single word.

Improvements ahead
- Make better UI in chatbox for "Active file" instead of just another tag in chatbox.

**Comment by charlwillia6** [comment]
@Fryingpannn @nang-dev @Himanshu-Singh-Chauhan - Do you want me to create an entire new PR when I fix the issues?

Second, what is your ides of this?

> Improvements ahead

>  Make better UI in chatbox for "Active file" instead of just another tag in chatbox.

**Pull request merged**